runtime/lua/vim/filetype.lua

```diff
@@ -2139,6 +2141,7 @@ local pattern = {
     ['%.git/worktrees/.*/config%.worktree$'] = 'gitconfig',
     ['/%.config/git/config$'] = 'gitconfig',
     ['/gitolite%-admin/conf/'] = starsetf('gitolite'),
+    ['/hypr/.*%.conf$'] = 'hyprlang',
```

Comment:
Just a note that these patterns with `.*` are really difficult to optimise around.

If the `.*` was replaced with something like `[^/]*` then we could do theoretically break up these patterns more easily and do quicker matches on the filename first. This would make generating a trie much easier.

I did a prototype of generating a trie, but I had to exclude too many patterns like this and the perf increase was only around 15%. If we can somehow simplify these patterns (remove `.*`), then the perf increase would be much more.

Is there any scope for this or must we copy these patterns from vim verbatim?

Reply:
Well, any deviation makes porting patterns harder, to the point where I will just stop. (Not blackmail; just projection.)

But from the point of view of the project, what matters is only the effective detection, so any optimization is fair game that does not affect that. (Unfortunately, the Vim tests tend to ignore the more complicated cases, so we can't rely on those for "safe" refactors.)

Reply:
Hmm ok. The main reason I mention is that I do think `.*` is generally not needed for almost all filetypes. Usually the convention will be something like `tool/file.conf` which then gets the pattern `tool/.*.conf`, even though `tool/a/b/c.conf` probably shouldn't be filetype matched.

Reply:
What is the actual proposed change here (in regex and pattern)?

Reply:
(Also, note how I fucked up even _this_ pattern...)

Reply:
Nothing concrete atm, I'm just trying to gauge what the design space is and how we can go about changing some of these patterns.

But mostly it'll be replacing `.*` with something more specific and adding more anchors (`^` and `$`). Both these things will enable breaking down patterns into multiple smaller patterns that can be structured and shared.

Reply:
The internal representation is completely open; the only constraint is matching observable Vim behavior and being able to reasonably port Vim filetypes.

Reply:
I'm suggesting that these patterns are tightened up so while `hypr/a.conf` would match as `hypr`, `hypr/a/b.conf` would not; so this would be an observable difference. My argument is that `hypr/a/b.conf` likely isn't valid so the observable vim behaviour is wrong (and slow).

Reply:
That would be a change that would have to go through Vim first. Be prepared for arguments.

(No argument from me, although I would expect that at least for some patterns, this _would_ be valid, so a case-by-case decision will be needed.)

Reply:
Ok sure, thanks for the advice. If I continue with this, then I'll take that route. I'll draft-PR what I've done so far at some point for reference and maybe make a tracking issue.

Reply:
In general, this sounds to me like a direction we would want to go in. But realistically, there would be significant friction from upstream, not least due to the many filetype maintainers that will have their say. It might actually pay off to rewrite their filetype.vim along the same lines as filetype.lua (but in vim9script, of course) and make these changes as part of the rewrite. (Not that I see any of us volunteering for that, nor being able to nerd-snipe any of the usual upstream suspects...)